File: 01161. Maximum Level Sum of a Binary Tree/solution.go
/**
 * Problem: 1161. Maximum Level Sum of a Binary Tree
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func maxLevelSum(root *TreeNode) int {
	// Initialize a queue for BFS traversal, starting with the root node
	queue := []*TreeNode{root}
	// Position pointer for the current node in the queue
	pos := 0
	// Marks the end index of the current level in the queue
	levelEnd := 0

	// Stores the maximum sum found so far, initialized with root's value
	maxSum := root.Val
	// Stores the level with the maximum sum, initialized to level 1
	maxLevel := 1

	// Accumulates the sum of node values at the current level
	levelSum := 0
	// Tracks the current level number, starting from 1
	currentLevel := 1

	// Iterate through the queue using BFS
	for pos < len(queue) {
		node := queue[pos]
		// Add the current node's value to the level sum
		levelSum += node.Val

		// Enqueue left child if it exists
		if node.Left != nil {
			queue = append(queue, node.Left)
		}
		// Enqueue right child if it exists
		if node.Right != nil {
			queue = append(queue, node.Right)
		}

		// If we've reached the end of the current level
		if pos == levelEnd {
			// Update maxSum and maxLevel if the current level's sum is greater
			if levelSum > maxSum {
				maxSum = levelSum
				maxLevel = currentLevel
			}

			// Move to the next level
			currentLevel++
			// Reset the sum for the next level
			levelSum = 0
			// Update the end index for the next level
			levelEnd = len(queue) - 1
		}

		// Move to the next node in the queue
		pos++
	}

	// Return the level with the maximum sum
	return maxLevel
}

File: 01161. Maximum Level Sum of a Binary Tree/solution.js
/**
 * Problem: 1161. Maximum Level Sum of a Binary Tree
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 3 ms (Beats 100%)
 */

/**
 * Returns the 1-indexed level with the highest sum in a binary tree
 *
 * @param {TreeNode} root - Root of the binary tree
 *
 * @returns {number} - Level with maximum sum
 */
const maxLevelSum = (root) => {
  // Initialize an array to store the sum of values at each level
  let levelSums = []

  // Define a helper function for BFS traversal to calculate level sums
  const traverseLevels = (currentNode, currentLevel) => {
    const { val, left, right } = currentNode // Destructure node properties

    // If this is the first node at the current level, initialize the sum
    if (levelSums.length < currentLevel + 1) levelSums.push(val)
    // Otherwise, add the node's value to the existing sum for the level
    else levelSums[currentLevel] += val

    // Recursively traverse the left subtree, increasing the level
    if (left) traverseLevels(left, currentLevel + 1)
    // Recursively traverse the right subtree, increasing the level
    if (right) traverseLevels(right, currentLevel + 1)
  }

  // Reset the levelSums array before traversal
  levelSums = []

  // Start BFS traversal from the root node at level 0
  traverseLevels(root, 0)

  // Initialize variables to track the maximum sum and its corresponding level index
  let maxSum = -Infinity
  let maxLevelIndex = 0

  // Iterate through the levelSums array to find the level with the maximum sum
  for (let i = 0; i < levelSums.length; i++) {
    if (levelSums[i] > maxSum) {
      maxSum = levelSums[i]
      maxLevelIndex = i
    }
  }

  // Return the 1-indexed level with the highest sum
  return maxLevelIndex + 1
}
